src/shared/sleep-config.c

```diff
@@ -102,6 +102,84 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
         return 0;
 }
 
+bool is_battery_low(void) {
+        _cleanup_free_ char *bat_cap_level = NULL;
+        int r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity_level", &bat_cap_level);
+        if (r < 0) {
+                log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity_level: %m");
+                return false;
+        }
+
+        log_debug("Current battery level: %s", bat_cap_level);
+
+        return STR_IN_SET(bat_cap_level, "Critical", "Low");
+}
+
+int read_battery_capacity(void) {
+        _cleanup_free_ char *bat_cap = NULL;
+        int battery_capacity, r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity", &bat_cap);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity: %m");
+
+        r = safe_atoi(bat_cap, &battery_capacity);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to convert battery capacity char value to int: %m");
+
+        log_debug("Current battery charge level: %d%%", battery_capacity);
+
+        return battery_capacity;
+}
+
+int get_battery_discharge_rate(void) {
+        _cleanup_free_ char *dirpath = NULL;
+        int stored_discharge_rate, r;
+        char *dis_rate;
+        sd_id128_t machine_id = SD_ID128_NULL;
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get machine-id: %m");
+
+        dirpath = path_join("/var/lib/systemd/sleep", SD_ID128_TO_STRING(machine_id), "battery_discharge_rate");
+        if (!dirpath)
+               return log_oom();
+
+        r = read_one_line_file(dirpath, &dis_rate);
+        if (r < 0)
+               return log_error_errno(r, "Failed to read discharge rate from %s: %m", dirpath);
+
+        r = safe_atoi(dis_rate, &stored_discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to convert discharge rate fetched from %s location from char value to int: %m", dirpath);
+
```

Comment:
safety checks should be done here - stored_discharge_rate should be greater than 0, and smaller than some arbitrary large limit, say 10000 - given it's a percentage-per-hour value, it should be something sensible (to avoid corruption, etc). Also there should be a comment in these functions saying that it's a percentage-per-hour value that is being saved.

Reply:
There is a possibility of the value to be 0. In that case it will not be written to file ?

Reply:
I think it makes sense to not update the value in file if it is not greater than 0 or less than 1000.

Reply:
if it's 0 it's either a calculation gone wrong, or it was on AC power, so either way it's not really a discharge rate - it can't be used to make assumptions

Reply:
right.